postgres/datadog_checks/postgres/statements.py

```diff
@@ -0,0 +1,191 @@
+            )
+            return
+
+        cursor = db.cursor(cursor_factory=psycopg2.extras.DictCursor)
+
+        rows = self._execute_query(
+            cursor,
+            STATEMENTS_QUERY.format(
+                cols=', '.join(self._pg_stat_statements_query_columns),
+                pg_stat_statements_view=self.config.pg_stat_statements_view,
+            ),
+        )
+        if not rows:
+            return
+
+        def row_keyfunc(row):
+            # old versions of pg_stat_statements don't have a query ID so fall back to the query string itself
+            queryid = row['queryid'] if 'queryid' in row else row['query']
+            return (queryid, row['datname'], row['rolname'])
+
+        rows = self._state.compute_derivative_rows(rows, PG_STAT_STATEMENTS_METRIC_COLUMNS.keys(), key=row_keyfunc)
+        metric_limits = (
+            self.config.statement_metric_limits if self.config.statement_metric_limits else DEFAULT_METRIC_LIMITS
+        )
+        rows = apply_row_limits(rows, metric_limits, 'calls', True, key=row_keyfunc)
+
+        for row in rows:
+            try:
+                normalized_query = datadog_agent.obfuscate_sql(row['query'])
+            except Exception as e:
+                logger.warning("Failed to obfuscate query '%s': %s", row['query'], e)
+                continue
+
+            # The APM resource hash will use the same query signature because the grouped query is close
+            # enough to the raw query that they will intersect frequently.
+            query_signature = compute_sql_signature(normalized_query)
+            apm_resource_hash = query_signature
+            tags = ['query_signature:' + query_signature, 'resource_hash:' + apm_resource_hash] + instance_tags
+            for column, tag_name in PG_STAT_STATEMENTS_TAG_COLUMNS.items():
+                if column not in row:
+                    continue
+                value = row[column]
+                if column == 'query':
+                    value = self._normalize_query_tag(value)
+                tags.append('{tag_name}:{value}'.format(tag_name=tag_name, value=value))
+
+            for column, metric_name in PG_STAT_STATEMENTS_METRIC_COLUMNS.items():
+                if column not in row:
+                    continue
+                value = row[column]
+                if column == 'total_time':
+                    # convert milliseconds to nanoseconds
+                    value = value * 1000000
+                instance.count(metric_name, value, tags=tags)
+
+    def _normalize_query_tag(self, query):
+        """Normalize the query value to be used as a tag"""
+        # Truncate to metrics tag limit
+        query = query.strip()[:200]
+        if self.config.escape_query_commas_hack:
+            # Substitute commas in the query with unicode commas. Temp hack to
+            # work around the bugs in arbitrary tag values on the backend.
+            query = query.replace(', ', '，').replace(',', '，')
```

Comment:
This seems like it could lead to very hard to find bugs.  Can we pick another explicit ASCII character or more obvious unicode character instead?

Reply:
Unfortunately, no, not one that we've found. For some context, this is a very hacky workaround to deal with the "arbitrary tag values" feature. There are many places throughout our stack that break when you try to submit tags whose values contain ascii commas. Even submitting a unicode comma is replaced on the backend with an ascii comma, which breaks dashboards and the query language.

Ideally, this hack exists for as short as possible, but there is work in Q4 for several metrics teams to allow us to get rid of this.

Reply:
I moved this to a shared util in `datadog_checks_base`. See the PR here: https://github.com/DataDog/integrations-core/pull/7860. If that's approved, I'll merge and update mysql/postgres PRs to use that function instead.